net/tcp/tcp_send_buffered.c

```diff
@@ -604,170 +603,116 @@ static uint16_t psock_send_eventhandler(FAR struct net_driver_s *dev,
 
   else if ((flags & TCP_REXMIT) != 0)
     {
-      rexmit = true;
-    }
-
-  if (rexmit)
-    {
+      size_t sndlen;
       FAR struct tcp_wrbuffer_s *wrb;
-      FAR sq_entry_t *entry;
+      bool rexmit_from_write_q = false;
 
-      ninfo("REXMIT: %04x\n", flags);
+      /* According to RFC 6298 (5.4), retransmit the earliest segment
+       * that has not been acknowledged by the TCP receiver.
+       */
 
-      /* If there is a partially sent write buffer at the head of the
-       * write_q?  Has anything been sent from that write buffer?
+      /* If the unacked_q is not empty, retrasmit the eariest segment
+       * that is in the head of the unacked_q.
        */
 
-      wrb = (FAR struct tcp_wrbuffer_s *)sq_peek(&conn->write_q);
-      ninfo("REXMIT: wrb=%p sent=%u\n", wrb, wrb ? TCP_WBSENT(wrb) : 0);
+      wrb = (FAR struct tcp_wrbuffer_s *)sq_peek(&conn->unacked_q);
```

Comment:
Have you tested the situation of continuous packet loss? I'm not sure if such a change can cover all scenarios. In the case of multiple packets being continuously discarded, it is more efficient to retransmit all of them, especially for wired networks.

Reply:
I have not tested continuous packet loss for this. As I can understand, a continuous packet loss is not a normal situation and it is subject to break the TCP connection.
Could you provide a link to RFC for retransmitting all the segments?

Reply:
Continuous packet loss (1~3) is common case on wireless device. The current stack does not support selective-ack, so I suggest that similar changes should be as cautious as possible,
https://datatracker.ietf.org/doc/html/rfc5681    3.2.3

Reply:
RFC 3.2.3 says "The lost segment starting at SND.UNA MUST be retransmitted ...". Only one segment must be retransmitted.
I was thinking you meant continuous packet loss say 30 seconds (all segments are lost inside of say 30 seconds).
If you mean only 1 to 3 lost segments, then yes, when I was testing the patch using packet loss simulation, there were cases with several lost segments (..., good, lost, about 5 good, lost, ...), and the new algorithm worked just fine (compared to the existing implementation that floods with excessive retransmissions).
However, I did not see cases (during the packet loss simulation) with 2 or 3 lost segments in series such as: ..., good, lost, lost, good, .... Are you concerned namely with the case?

Reply:
Yes, I don't deny the benefits of retransmitting the earliest packets, but if you can keep all retransmissions during the timer callback, I think it's more reasonable

Reply:
I am just saying that no RFC (at least I was not able to find any) says that in some case all the segments must (or at least should) be retransmitted.
Retransmitting all the segments makes things much worse (flooding with retransmitted packets congesting the channel even more).
I am sure we should not introduce anything new from ourselves, instead we should strictly follow the RFCs.

Reply:
I agree with you, but your change will makes the situation worse in the case of continuous packet loss, this enhancement needs to be based on selective-ack. Without this feature, sending more is better than sending less

Reply:
> but your change will makes the situation worse in the case of continuous packet loss

Do you mean namely the following pattern?
..., good, lost, lost, good, ...

Reply:
Yes

Reply:
Good. I will try to emulate namely that pattern and see how the stack behaves.

Reply:
I was able to emulate a loss of two segments in a row as follows:
```
$ sudo iptables -A INPUT -p tcp --dport 5471 -m statistic --mode nth --packet 10 --every 1000000 -j DROP
$ sudo iptables -A INPUT -p tcp --dport 5471 -m statistic --mode nth --packet 10 --every 1000000 -j DROP

```
I have tested tcp_send_unbuffered and tcp_sendfile for the case. These both modes perform fine:
![tcp_send_unbuffered](https://user-images.githubusercontent.com/10873263/155453898-e5c2d010-0c7a-4632-ad38-b0502592d14f.png)

Reply:
Also I have tested tcp_send_buffered in case of two lost segments in a row.
Before PR #5589 the result is the following:
![tcp_send_buffered_rfc_2001_flooding](https://user-images.githubusercontent.com/10873263/155455626-b0a2bd8b-2097-4dba-81a2-6f48eeb81ae9.png)

With applied PR #5589 the result is the following:
![tcp_send_buffered_PR5589_rfc_2001](https://user-images.githubusercontent.com/10873263/155455070-3377014c-1626-4c22-847c-dba60babb70f.png)

Reply:
> I agree with you, but your change will makes the situation worse in the case of continuous packet loss

As you can see on the screenshot above, the root cause of the exposed issue is that the existing Fast Retransmit algorithm (#2414) normally detects only the first lost segment, however it does not detect the second lost segment (the detector does not recover to the initial state).
I have fixed the issue of the Fast Retransmit algorithm and upgraded it from RFC 2001 to 5681 (PR #5598).
Now if both PRs #5589 and #5598 are applied, the result is the following (for the case of two lost segments in a row):

![tcp_send_buffered_PR5589_rfc_5681](https://user-images.githubusercontent.com/10873263/155456717-5ee1888d-1f3a-4f61-b5e4-4927e480d257.png)

Reply:
see my comment on #5598